snapcraft/internal/elf.py

```diff
@@ -175,6 +176,17 @@ def _extract(self, path):  # noqa: C901
         with open(path, 'rb') as fp:
             elf = elftools.elf.elffile.ELFFile(fp)
 
+            # A set of fields to identify the architecture of the ELF file:
+            #  EI_CLASS: 32/64 bit (e.g. amd64 vs. x32)
+            #  EI_DATA: byte orer (e.g. ppc64 vs. ppc64le)
+            #  e_machine: instruction set (e.g. x86-64 vs. arm64)
+            #
+            # For amd64 binaries, this will evaluate to:
+            #   ('ELFCLASS64', 'ELFDATA2LSB', 'EM_X86_64')
```

Comment:
Do we need all this information? Or can we just use `get_machine_arch()`?

Reply:
The point of this code is to handle the case where a snap contains ELF files for multiple architectures.  Some examples where this happen include:

1. the core snap for amd64/arm64, which contains libraries to support the matching 32-bit architecture:

    /snap/core/current/lib/i386-linux-gnu/libc-2.23.so
    /snap/core/current/lib/x86_64-linux-gnu/libc-2.23.so

2. architecture all snaps that contain executables for multiple architectures.

The aim here is to e.g. make sure a 32-bit executable depending on libfoo.so is matched to the 32-bit version of that library, and the equivalent for 64-bit.  The `platform.get_machine_arch` function doesn't help with this.

Reply:
bah.  I was looking at the wrong function.  So, `ELFFile.get_machine_arch()` is not sufficient because it only references `e_machine`:

https://github.com/eliben/pyelftools/blob/master/elftools/elf/elffile.py#L212

So it can't distinguish amd64 vs. x32 binaries (both share the same e_machine value, differ in word size), or ppc64 vs. ppc64le (both share same e_machine and word size, but have different byte order).  Also, get_machine_arch() folds everything but 5 values to "<unknown>".